### PR TITLE
Fix Rosetta Bug

### PR DESCRIPTION
**Pull request by LuttyYang**

### CODE REVIEW

**Review by LeoHChen — COMMENTED** [review]
A few comments. This is a big PR, but I would need more context so that we know why this PR is needed.

1) Please get @peekpi to review the evm change and tracer change
2) need more detailed descriptions of all the fixes
3) please also add all the errors log encountered during the test into a github issue and refer it in this PR

**Comment by LeoHChen** [comment]
Also, it failed travis CI. please fix it.

**Review by jhd2best — APPROVED** [review]

**Comment by peekpi** [comment]
I checked the changes except for rosetta/*, looks good to me.

**Comment by jhd2best** [comment]
errors log encountered during the test related this PR.  https://docs.google.com/document/d/1gRvaZL1hTP7TNxrrzEnZUfckKiJxMT1muyjy1GJeP4w

**Comment by LeoHChen** [comment]
https://app.travis-ci.com/github/harmony-one/harmony/jobs/561034274

Travis CI failed on Rosetta checker, please take a look. @jhd2best

**Review by LeoHChen — APPROVED** [review]

**Comment by LeoHChen** [comment]
approved but please fix the few comments and travis test.

File: rosetta/services/tx_operation.go [hunk 1]
```diff
@@ -454,17 +455,19 @@ var (
 	// internalNativeTransferEvmOps are the EVM operations that can execute a native transfer
 	// where the sender is a contract address. This is also known as ops for an 'internal' transaction.
 	// All operations have at least 7 elements on the stack when executed.
-	internalNativeTransferEvmOps = map[string]interface{}{
-		vm.CALL.String():         struct{}{},
-		vm.CALLCODE.String():     struct{}{},
-		vm.SELFDESTRUCT.String(): struct{}{},
+	internalNativeTransferEvmOps = map[vm.OpCode]interface{}{
```
[the remark this answers is not in the document]
Comment (LuttyYang) [reply rosetta/services/tx_operation.go]: This code is not needed now, it is currently implemented using the same trace as rpc

File: core/vm/contracts_write.go [hunk 2]
```diff
@@ -112,19 +112,25 @@ func (c *stakingPrecompile) RunWriteCapable(
 	if err != nil {
 		return nil, err
 	}
+
+	var rosettaBlockTracer RosettaTracer
+	if tmpTracker, ok := evm.vmConfig.Tracer.(RosettaTracer); ok {
+		rosettaBlockTracer = tmpTracker
+	}
```
[the remark this answers is not in the document]
Comment (LuttyYang) [reply core/vm/contracts_write.go:119]: No, there will be multiple callers, only the implementation will be returned when rosetta is called

File: core/evm.go [hunk 3]
```diff
@@ -47,6 +47,8 @@ type ChainContext interface {
 	// ReadDelegationsByDelegator returns the validators list of a delegator
 	ReadDelegationsByDelegator(common.Address) (stakingTypes.DelegationIndexes, error)
 
+	ReadDelegationsByDelegatorAt(delegator common.Address, blockNum *big.Int) (m stakingTypes.DelegationIndexes, err error)
```
[the remark this answers is not in the document]
Comment (LuttyYang) [reply core/evm.go:51]: added

**Comment by jhd2best** [comment]
@LeoHChen all travis tests passed. we can merge it.

**Pull request merged**